### DISCUSSION TITLE
main.py does not exist!

### DISCUSSION THREAD

**Discussion by hcgrant** [micropython/micropython#11839] [RP2040 / Pico] [answered]
I have a number of picows round the house measuring temp and humidity. I have written some code to put in the temp/humid monitors to enable me to update/modify the code in the picows.  My PC sends UDP packets to the picow to perform various actions. The following commands work ok ..List Files (List the files in the picow), Load File  (copies a file to the picow), Delete File and Reset Pico.
One command "Get File" (Copies the file from the picow to my PC) works ok apart from when I try to "get" the main.py file.  This also happens when using Thonny and the running program is main.py or some other file name.  The code in the picow responding to the Get File command is ..  
 ```
def get_file(filename):                
        print("getting "+filename)
        try :
            fl = open(filename,"rt")
            msg = fl.read()
            send_message(msg)
            fl.close()
            print(" Sent contents of "+filename)
        except : 
            msg = (filename+" does not exist")
            send_message(msg)
        
```
I suspect that I dont quite understand how the picow works , any suggestion most welcome

**Comment by peterhinch** [comment]
There should be no problem reading the running file: by the time the code starts to run it has been compiled to bytecode which is located in RAM. The "running" file is just sitting in the filesystem being ignored.

I don't use Thonny: is it possible that your tools precompile the file to an `mpy` file, copying that to the Pico W? You might want to use a tool like `mpremote` to check whether `main.py` is actually in the filesystem.

**Comment by karfas** [comment]
Your code:
```
       try :
           fl = open(filename,"rt")
           msg = fl.read()
           send_message(msg)
           fl.close()
           print(" Sent contents of "+filename)
       except : 
           msg = (filename+" does not exist")
```

Most likely your exception handling is way too broad.
You will get the `filename+" does not exist" ` message as well when
- `fl.read()` fails, maybe due to out-of-memory conditions
- `send_message()` fails, maybe due to out-of-memory conditions

I would print the real exception in the except block.

**Comment by hcgrant** [comment]
Many thanks for both your suggestions .. Ive had a brief look at mpremote and I think its worth  spending more time to get used to it.
Ive followed karfas' advice and found that the problem is a failure in send_message(msg). I think the problem is the size of main.py. A smaller config file can be "got". I need to find out the limit of UDP packet size. The len(msg) value for the smaller config file is 398 whereas the value for main.py is 10193.

**Comment by hcgrant** [comment, marked as answer]
Ive written a small program to send ever increasing UDP messages until I get an error .. ENOMEM  This value = 7742 bytes.
On my desktop this value is 63552. 
Micropython has obviously changed the the max message size since picos etc are a bit smaller than my desktop  :-)))

**Reply by karfas** [reply]
I doubt there is a "max message size" defined somewhere in micropython (leaving alone the ~64kB limit for UDP packages defined by the UDP protocol).

Your desktop can allocate many 64kB buffers out of several GB, whereas the PICO as well as many other embedded systems have maybe 100-150kB available (minus anything your program uses, minus memory fragmentation).

**Comment by kingosapplerepair** [comment]
Script file 'main.py' does not exist.

**Reply by robert-hh** [reply]
Yes, that is the title of this post. Why do you repet it?

**Reply by kingosapplerepair** [reply]
Thank you for this attention! 
I'm sorry for this mistake, I m trying to convert a script python to executable file  and this error (Script file 'main.py' does not exist.)

**Comment by kingosapplerepair** [comment]
Thank you for this attention!
I'm sorry for this mistake, I m trying to convert a script python to
executable file
